spyder_kernels/customize/namespace_manager.py

```diff
@@ -52,6 +57,21 @@ def __enter__(self):
         if '__file__' in self.namespace:
             self._previous_filename = self.namespace['__file__']
         self.namespace['__file__'] = self.filename
+        if (self._file_code is not None
+                and not PY2
+                and isinstance(self._file_code, bytes)):
+            try:
+                self._file_code = self._file_code.decode()
+            except UnicodeDecodeError:
+                # Setting the cache is not supported for non utf-8 files
+                self._file_code = None
+        if self._file_code is not None:
+            # '\n' is used instead of the native line endings. (see linecache)
+            # mtime is set to None to avoid a cache update.
+            linecache.cache[self.filename] = (
```

Comment:
`linecache` is what `pdb` and `traceback` are using. So I am adding code here for any package using `linecache` to use.

Reply:
This PR isn't really about caching code in Spyder. Perhaps the title should be changed to "Runfile to retrieve code from the Spyder editor by default" The linecaching is just a way to make it look like a real file to python.

I suggest the change in title also because a execution cache will also have to be made at some point for a queue implementation, although this PR makes it possible to execute code file from a cache in the kernel.

Reply:
Well for files that are not open in spyder, this will only cache the code. Maybe `Runfile to retrieve code from the Spyder editor and add to linecache` would be more appropriate.